### ISSUE TITLE
Suggestions to improve/correct the records proposal

### ISSUE THREAD

**Issue by RexJaeschke** [dotnet/docs#28648]
Using this proposal, I’ve been writing a formal spec for this feature to add to the C# standard, and have the following comments:

## Comment 1: “parameter_list grammar rule is not defined”

The proposal contains the following grammar:

> ```ANTLR
> record_declaration
>     : attributes? class_modifier* 'partial'? 'record' identifier type_parameter_list?
>       parameter_list? record_base? type_parameter_constraints_clause* record_body
>     ;
> ```

which refers to the grammar rule *parameter_list*.

As of the V6 C# Standard, no such rule is defined, and none is defined in the proposals for V7, V8, or V9 either. (However, I do see other references to this rule in the V10 proposal “[record-structs.md](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/record-structs.md)” and in the active proposal, “[non-record-primary-constructors.md](https://github.com/dotnet/csharplang/blob/main/proposals/non-record-primary-constructors.md).”)

So, what is the definition of this rule, and when/where was it defined in terms of proposals and product releases? Or should it be defined as part of this proposal?

Here’s my attempt at defining this rule, assuming it behaves *exactly* like that for a method (which it may well not):

```ANTLR
parameter_list
    : '(' formal_parameter_list? ')'
    ;
```

where *formal_parameter_list* is the rule we currently use to declare the parameters of a method.

Now, I did find in the “rejected” proposals folder, proposal “[records.md](https://github.com/dotnet/csharplang/blob/main/proposals/rejected/records.md).” In this earlier incarnation of a records proposal, it shows the following in the section titled, “record type declarations,” which contains an equivalent notion of *parameter_list* but calls it *record_parameters* instead:

> The syntax for a `class` or `struct` declaration is extended to support value parameters; the parameters become properties of the type:
> ```antlr
> class_declaration
>     : attributes? class_modifiers? 'partial'? 'class' identifier type_parameter_list?
>       record_parameters? record_class_base? type_parameter_constraints_clauses? > class_body
>     ;
> …
> record_parameters
>     : '(' record_parameter_list? ')'
>     ;
> 
> record_parameter_list
>     : record_parameter
>     | record_parameter ',' record_parameter_list
>     ;
> 
> record_parameter
>     : attributes? type identifier record_property_name? default_argument?
>     ;
> 
> record_property_name
>     : ':' identifier
>     ;
> …
> ```

However, none of this grammar made it into the final proposal.

## Comment 2: “Examples of synthesized equality members”

In the section titled, “Equality members”, we have the following:

> For example, consider the following record types:
> ```C#
> record R1(T1 P1);
> record R2(T1 P1, T2 P2) : R1(P1);
> record R3(T1 P1, T2 P2, T3 P3) : R2(P1, P2);
> ```
> For those record types, the synthesized equality members would be something like:
> ```C#
> class R1 : IEquatable<R1>
> {
> …
> }
> class R2 : R1, IEquatable<R2>
> {
> …
> }
> class R3 : R2, IEquatable<R3>
> {
> …
> }
> ```

Shouldn’t the three occurrences of `class` be `record`?

Similarly, in the section titled, “Printing members: PrintMembers and ToString methods”, we have:

> For example, consider the following record types:
> ``` csharp
> record R1(T1 P1);
> record R2(T1 P1, T2 P2, T3 P3) : R1(P1);
> ```
> For those record types, the synthesized printing members would be something like:
> ```C#
> class R1 : IEquatable<R1>
> {
> …
> }
> class R2 : R1, IEquatable<R2>
> {
> …
> }
> ```

Shouldn’t the two occurrences of `class` be `record`?

## Comment 3: “Generic constraints”

The proposal says nothing about these in the context of records, and it should.

From my tests, although there is no generic constraint that requires a type to be a record, a record type can satisfy either the `class` or `class?` constraint. And, a record type can also satisfy a specific base record type constraint (currently known as a base class constraint).

## Comment 4: “record_base’s argument_list”

The proposal contains the following grammar:

> ```antlr
> record_base
>     : ':' class_type argument_list?
>     | ':' interface_type_list
>     | ':' class_type argument_list? ',' interface_type_list
>     ;
> ```

Given that the current (draft V6) C# standard has

```ANTLR
argument_list
    : argument (',' argument)*
    ;
```

the two occurrences of `argument_list?` are incomplete! Where they are used, the compiler allows one to have nothing (supported by the proposal), empty parens (not supported by the proposal), or an argument list inside parens (not supported by the proposal). What I think is needed instead is

```antlr
record_base
    : ':' class_type ('(' argument_list? ')')?
    | ':' interface_type_list
    | ':' class_type ('(' argument_list? ')')? ',' interface_type_list
    ;
```

However, that could be made more readable by having the following:

```antlr
record_base
    : ':' class_type record_base_arguments?
    | ':' interface_type_list
    | ':' class_type record_base_arguments? ',' interface_type_list
    ;

record_base_arguments
    : '(' argument_list? ')'
    ;
```

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 9aba70b5-ee1a-90fd-b3ed-3384cd9f6512
* Version Independent ID: ab712864-bbc0-e4f9-cb32-15e12cd82c65
* Content: [Records - C# 9.0 draft specifications](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/records)
* Content Source: [proposals/csharp-9.0/records.md](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-9.0/records.md)
* Product: **dotnet-csharp**
* Technology: **csharp-spec**
* GitHub Login: @dotnet-bot
* Microsoft Alias: **wiwagn**

**Comment by BillWagner** [comment]
Ping @jcouv for who this should get routed to.

My first thought is that Comments 1, 3, and 4 are valid. I think the test referenced in Comment 2 is meant to describe the lowered type declaration created by the compiler, so `class` would be correct.

**Comment by BillWagner** [comment]
@RexJaeschke Are there updates to the comments based on the discussion with @MadsTorgersen on the strategy for describing records?

**Comment by RexJaeschke** [comment]
@BillWagner, Now that I am much wiser about records, I withdraw Comment 2. In a separate comment below, I'll explain the latest thinking about records by people on the C# standard's committee, but that doesn't change the core issues I raise in the other three Comments.

**Comment by RexJaeschke** [comment]
The current records proposal (and on-line MS tutorial material) presents a record as a type, *separate* from `class`. And that's how I started spec'ing it. However, once I got a ways into it, that approach looked like having a significant (negative) impact on the C# spec, especially when one considered the V10 feature record struct as yet another type. After some discussion among several C# standards committee members, @MadsTorgersen proposed that `record` simply be an optional characteristic of a class. And that is how I'm now spec'ing it. (That approach is proving to be *much* simpler, and doesn't affect the feature's syntax or semantics, or contradict the implementation.)

To make this clear, here's the grammar from my revised approach:

```ANTLR
class_declaration
  : attributes? class_modifier* 'partial'? [class_tag]() 'class' identifier type_parameter_list?
  parameter_list? class_base? type_parameter_constraints_clause* class_body ';'?
  ;

class_tag
  : 'class'            // V9
  : 'record'? 'class'    // changed to this in V10
  | 'record'
  ;

class_base
  : ':' class_type base_argument_list?
  | ':' interface_type_list
  | ':' class_type base_argument_list? ',' interface_type_list
  ;

base_argument_list
    : '(' argument_list? ')'
    ;

class_body
  : '{' class_member_declaration* '}' ';'?
  | ';'
  ;
```

In the (hopefully) few places in which we need to distinguish, we can define the following terms in the Classes chapter:
If the *class_tag* in a *class_declaration* contains `record`, that class is a ***record class***; otherwise, that class is a ***non-record class***.

This means that we will *not* use the word "record" on its own; it will always be "record class," in anticipation of eventually having V10's "record struct" as well, and in V10, allowing `record class`.

**Comment by BillWagner** [comment]
@RexJaeschke I think Comment 3 can also be withdrawn. If we spec `record` as an *optional characteristic of a class*, if follows that a `record` (or `record class`) satisfies the `class` constraint.

Do you agree?

**Comment by RexJaeschke** [comment]
@BillWagner Yes, I withdraw Comment 3.

**Comment by BillWagner** [comment]
After reviews on dotnet/csharplang#6157, closing without action.

**Comment 1**: See https://github.com/dotnet/csharplang/pull/6157#discussion_r880915496.  Because these are legal, this should not be changed:

```csharp
public record R();
```

The argument list for a positional record can be empty. That means we can use the same `parameter_list` production for method parameter lists. There's no reason for a new production.

**Comment 4**: This is also legal:

```csharp
public record R
{
   public int F { get; init; }
}

public record D : R() // note empty argument list
{
    public string F2 { get; init; }
}
```

That means the existing grammar for the `record_base` production is correct.

**Comment by RexJaeschke** [comment]
@BillWagner I'm confused! Certainly the MS proposal, as published, is *not* what is being proposed by me to the Ecma standards committee. 

Re Comment 1, while I do use the grammar rule *parameter_list* in my grammar rewrite, I have defined it, as it does not currently exist and it is *not* defined by the MS proposal, and we cannot simply replace it with *formal_paramerer_list*.

Re Comment 4, this grammar has also been rewritten in my proposal, and it does handle the omission I pointed out.

The bottom line then is, "Should the MS proposal be revised to reflect the new thinking?" If not, it likely will be misleading to anyone who reads it.

**Comment by BillWagner** [comment]
@RexJaeschke That's a good question. I've got an internal thread going because the compiler team has an [ANTLR grammar file](https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Generated/CSharp.Generated.g4) for the implementation. It's generated from an [internal XML representation](https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Syntax/Syntax.xml). 

The `parameter_list` production in the implementation grammar is *similar to* our [`formal_parameter_list`](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/classes#14621-general) production. (the difference is that the `parameter_list` doesn't include `params` arguments.

The standard's grammar and the implementation grammar have diverged. That creates some questions going forward. Does that matter? They serve different purposes. Will it add too much confusion? I'm not sure yet.

And, if we want the generated grammar to be *the same*, should we consider some way to merge tools? 

For all of these questions, I'm not sure yet. What are your thoughts?

**Comment by RexJaeschke** [comment]
@BillWagner Thanks for the pointer to the internally generated grammar. As a result, I plan to take a look at all the new grammar I added to the V7, 8, and 9 specs thus far, comparing it with that from the generated grammar to confirm I have covered all the bases and that I use equivalent rules.

The MS grammar uses *parameter_list*, which includes the delimiting `(...)`. As that precluded its use in indexers, the grammar has another rule, *bracketed_parameter_list*, which includes the delimiting `[...]`. The spec uses *formal_parameter_list* in all cases, and then adds on the `(...)` or `[...]` delimiters, as appropriate.

You wrote that *parameter_list* doesn't include *params* arguments. In that regard, it certainly doesn't use the same  approach as the spec, but I see the use of a special *literal_expression* token `__arglist`, which looks like that might serve the same purpose.

I also noticed the MS grammar has separate rules for *class_declaration* and *record_declaration*, just like I did until Mads suggested we make a record a variant of a class.

Re any sort of merging, here are my immediate thoughts:

1. For openness, we've moved the spec to a public repo. More eyes on are better!
2. MS no longer maintains its own spec version, and now points to the Ecma version. One source of truth is good!
3. The more resources and tools than can be shared between implementation and specification, the better. More efficient, cost-effective.
4. It is not uncommon for formal specs to point to, or even provide, a reference implementation. (The Java spec has one.) Given the status of Roslyn, might that reasonably be considered C#'s reference implementation?
5. If we were to harmonize grammar and tooling, how would we handle things like the above-mentioned *parameter_list* vs. *formal_parameter_list* approaches? My initial reaction would be to use the approach from the reference implementation, unless we had good reason to not do so, **and** the implementation would be willing to change to our approach.
6. I haven't looked in detail at the generated grammar, but I'm wondering if it has left-recursive rules, like we still have, but for which Nigel provided an ANTLR-validation workaround. If so, then how do tools processing the MS grammar deal with that? If it does not, presumably the MS grammar is *not* fed into an ANTLR processor. In any event, merging these kinds of approaches could involve some work, and Nigel would have an opinion re this.
7. Although we tend to not think about language extensions above and beyond the standard, I'm thinking that there might be good reason for the MS grammar to be a superset of the spec grammar, but whether it could be a strict superset, is another matter.
